### PR TITLE
test(rs-dapi): improve test coverage

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented

Improve unit test coverage for the `packages/rs-dapi/src/` module, which had only 39% line coverage.

## What was done?

Added comprehensive unit tests across 9 source files, raising overall line coverage from 39% to 52%. Tests focus on error conversion logic, JSON-RPC/gRPC protocol translation, deserialization helpers, metrics infrastructure, logging middleware, and error mapping -- all pure logic functions that are unit-testable without network dependencies.

**Coverage improvements by file:**

| File | Before | After | Delta |
|------|--------|-------|-------|
| `error.rs` | 7% | 94% | +87% |
| `protocol/jsonrpc_translator/error.rs` | 30% | 100% | +70% |
| `config/utils.rs` | 18% | 88% | +70% |
| `metrics.rs` | 33% | 87% | +55% |
| `error_mapping.rs` | 51% | 91% | +40% |
| `utils.rs` | 53% | 93% | +40% |
| `logging/middleware.rs` | 41% | 73% | +32% |
| `wait_for_state_transition_result.rs` | 0% | 19% | +19% |
| `logging/access_log.rs` | 80% | 90% | +10% |

**What is tested:**
- `error.rs`: `to_status`, `into_legacy_status`, `From` impls (boxed errors, dashcore_rpc errors, tonic::Status), all constructor helpers, `no_valid_tx_proof` (32-byte hash vs hashing), `map_join_result`, `MapToDapiResult`
- `jsonrpc_translator/error.rs`: `map_error` for all `DapiError` variants; `map_status` with all gRPC codes (both with messages and empty-message defaults)
- `utils.rs`: `generate_jsonrpc_id` uniqueness/format, `deserialize_string_or_number` from string/u64/i64/bool, `deserialize_to_string` from all types, `deserialize_string_number_or_null` including null/array error
- `metrics.rs`: `clamp_to_i64` boundary, `endpoint_label` for gRPC/JSON-RPC/HTTP with/without hints, `MethodLabel`/`Metric`/`Label`/`Outcome` enums, `Metrics` struct, all metric recording functions
- `config/utils.rs`: `from_str_or_bool` (true/false/1/0/yes/no/on/off + native bool + invalid), `from_str_or_number`
- `error_mapping.rs`: `map_tenderdash_message` (all 6 match arms + no-match), `TenderdashStatus::grpc_code` (standard + consensus code ranges), `grpc_message` fallbacks, `From<Value>` (object/non-object/missing-code/empty-message), Debug formatting, `base64_decode`, `From<TenderdashStatus> for StateTransitionBroadcastError`
- `logging/middleware.rs`: `detect_protocol_type` (JSON-RPC, gRPC content-type, gRPC headers, HTTP/2 paths, plain HTTP), `parse_grpc_path` (edge cases), `http_status_to_grpc_status` (all known + unknown codes), `extract_grpc_status` fallbacks
- `logging/access_log.rs`: `grpc_status_to_http_status` (remaining codes), gRPC JSON fields, null optional fields, combined format defaults
- `wait_for_state_transition_result.rs`: `build_wait_for_state_transition_error_response` (TenderdashClientError, generic DapiError, empty status message)

## How Has This Been Tested?

- `cargo test --package rs-dapi --all-features` -- all 267 tests pass (177 new)
- `cargo llvm-cov test --package rs-dapi --all-features` -- verified coverage improvements
- `cargo fmt --package rs-dapi` -- formatting verified

## Breaking Changes

None. Only test code was added.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

## Summary by CodeRabbit

* **New Features**
  * Improved error mapping to provide clearer translated error responses from external RPC sources.

* **Tests**
  * Extensive new unit tests across error handling, logging, metrics, configuration deserializers, protocol translation, platform services, and utilities to boost reliability and coverage.

### CODE REVIEW

File: packages/rs-dapi/src/error.rs [hunk 1]
```diff
@@ -295,6 +295,452 @@ impl<T: Sized, E: Into<DapiError>> MapToDapiResult<T> for Result<T, E> {
...
+
+    #[test]
+    fn from_dashcore_rpc_unknown_code_maps_to_unavailable() {
+        let rpc_err = dashcore_rpc::Error::JsonRpc(jsonrpc::Error::Rpc(jsonrpc::error::RpcError {
+            code: -999,
+            message: "unknown".to_string(),
+            data: None,
+        }));
+        let err: DapiError = rpc_err.into();
+        assert!(matches!(err, DapiError::Unavailable(msg) if msg.contains("-999")));
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dapi/src/error.rs:772]: Good catch — this is a pre-existing bug in the production error mapping code, not a test issue. The tests correctly document the current (incorrect) behavior. Opening an issue to fix the RPC code mapping separately.

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dapi/src/error.rs:772]: Created fix PR #3316 to correct the mapping. The tests in this PR will need updating after that merges — -8 should assert InvalidArgument and -1 should assert Internal.

**Review by QuantumExplorer — COMMENTED** [review]
## Audit Review

Solid test coverage across error mapping, deserialization helpers, logging, metrics, and protocol translation. All tests call real production functions (no fake/inline tests). The error mapping coverage is particularly thorough.

### Good practices
- Parameterized iteration where appropriate (`from_str_or_bool_true_values` iterates 7 inputs)
- Good boundary testing (`clamp_to_i64`, `no_valid_tx_proof` 32-byte vs non-32-byte paths)
- Async tests properly use `#[tokio::test]`
- `map_tenderdash_message` covers all 6 patterns + None return
- JSON-RPC empty-message default tests catch real edge cases

### Production code observations (pre-existing, not introduced by this PR)
1. **`detect_protocol_type` false positive**: Checks `req.headers().contains_key("te")` without inspecting value. Any HTTP request with `TE: gzip` gets misclassified as gRPC. Only `te: trailers` is gRPC-specific.
2. **JSON-RPC `data` field leaks internal details**: The `map_error` fallthrough puts `error.to_string()` in the `data` field, which can expose file paths, network addresses from `Io`, `Transport`, `Configuration` errors.
3. **`DapiError::InvalidRequest` maps to `Internal` in gRPC**: Not matched in either `to_status` or `into_legacy_status`, falls through to `Internal`. Arguably should be `InvalidArgument`.
4. **`Timeout` asymmetry**: `to_status` maps Timeout to `Internal`, but `into_legacy_status` maps it to `DeadlineExceeded`. `build_wait_for_state_transition_error_response` uses `to_status`, so timeouts get `Internal` code there.

File: packages/rs-dapi/src/error.rs [hunk 2]
```diff
@@ -295,6 +295,452 @@ impl<T: Sized, E: Into<DapiError>> MapToDapiResult<T> for Result<T, E> {
...
+        }
+    }
+
+    // -- map_join_result --
+
+    #[tokio::test]
+    async fn map_join_result_ok_ok() {
+        let handle = tokio::spawn(async { Ok::<_, DapiError>(42) });
+        let result = DapiError::map_join_result(handle.await);
+        assert_eq!(result.unwrap(), 42);
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/error.rs]: **Low: 36 bare `.unwrap()` calls across PR**

This PR introduces ~36 bare `.unwrap()` calls across test files. When a test fails at an unwrap, the panic gives no context. Use `.expect("descriptive message")` instead.

Notably, `from_str_or_bool_true_values` in `config/utils.rs` correctly uses `.unwrap_or_else(|_| panic!("failed for {}", input))` — proving the pattern is known. The inconsistency suggests these were rushed.

File: packages/rs-dapi/src/error.rs [hunk 3]
```diff
@@ -295,6 +295,452 @@ impl<T: Sized, E: Into<DapiError>> MapToDapiResult<T> for Result<T, E> {
...
+                assert_eq!(status.code, 42);
+            }
+            other => panic!("expected TenderdashClientError, got {:?}", other),
+        }
+    }
+
+    // -- constructor helpers --
+
+    #[test]
+    fn constructor_helpers_create_correct_variants() {
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/error.rs]: **Low: 5 assertions packed into one test**

This test validates 5 different constructor functions (`configuration`, `streaming_service`, `client`, `server`, `server_unavailable`). If the first assertion fails, the rest never execute. Consider splitting into separate tests, or at minimum add a comment explaining the grouping.

File: packages/rs-dapi/src/metrics.rs [hunk 4]
```diff
@@ -567,3 +567,189 @@ pub fn workers_active_inc() {
...
+    #[test]
+    fn gather_prometheus_returns_non_empty() {
+        let (buffer, content_type) = gather_prometheus();
+        // Buffer may be empty if no metrics have been touched yet, but should not panic
+        assert!(!content_type.is_empty());
+        let _ = buffer; // just ensure it doesn't panic
+    }
+
+    #[test]
+    fn cache_metrics_functions_do_not_panic() {
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/metrics.rs:601]: **Informational: Smoke tests — consider adding comments**

These `*_do_not_panic` tests only verify functions don't crash. They're more useful than they appear (Prometheus panics on label cardinality mismatches, double-registration, etc.), but a `// smoke test: verifies metric registration and label cardinality` comment would clarify their intent for future readers.

Same applies to `metrics_layer_new` (line 750) and `gather_prometheus_returns_non_empty` (line 589).

File: packages/rs-dapi/src/logging/middleware.rs [hunk 5]
```diff
@@ -403,4 +403,160 @@ mod tests {
...
+    fn detect_protocol_type_grpc_encoding_header() {
+        let req = Request::builder()
+            .header("grpc-encoding", "gzip")
+            .body(())
+            .unwrap();
+        assert_eq!(detect_protocol_type(&req), "gRPC");
+    }
+
+    #[test]
+    fn detect_protocol_type_te_header() {
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/logging/middleware.rs:482]: **Medium: Test passes but masks a production false positive**

This test correctly verifies `te: trailers` → gRPC. However, the production code at line ~216 checks `req.headers().contains_key("te")` without inspecting the value. A plain HTTP request with `TE: gzip` or `TE: chunked` would be misclassified as gRPC.

Consider adding a test documenting this behavior:
```rust
#[test]
fn detect_protocol_type_te_non_trailers_still_detected_as_grpc() {
    let req = Request::builder().header("te", "gzip").body(()).unwrap();
    // NOTE: This is arguably a false positive in production code
    assert_eq!(detect_protocol_type(&req), "gRPC");
}
```

File: packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs [hunk 6]
```diff
@@ -68,3 +68,236 @@ fn map_status(status: &dapi_grpc::tonic::Status) -> (i32, String, Option<Value>)
...
+
+    #[test]
+    fn map_error_not_found() {
+        let (code, msg, _) = map_error(&DapiError::NotFound("missing".into()));
+        assert_eq!(code, -32602); // ERR_NOT_FOUND
+        assert_eq!(msg, "missing");
+    }
+
+    #[test]
+    fn map_error_fallthrough_returns_internal() {
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs:145]: **Medium (production): Fallthrough leaks internal error details**

The test correctly verifies `data.is_some()` for the fallthrough, but the production code puts `error.to_string()` into the JSON-RPC `data` field. For errors like `Io`, `Transport`, `Configuration`, this can expose file paths, network addresses, or other system information to external callers.

*Pre-existing issue, not introduced by this PR. The test accurately documents current behavior.*

File: packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs [hunk 7]
```diff
@@ -310,3 +310,84 @@ pub(super) fn build_wait_for_state_transition_error_response(
...
+
+    #[test]
+    fn build_error_response_from_tenderdash_client_error() {
+        let tenderdash_status = TenderdashStatus::new(42, Some("td error".to_string()), None);
+        let error = DapiError::TenderdashClientError(tenderdash_status);
+        let response = build_wait_for_state_transition_error_response(&error);
+        let inner = response.into_inner();
+        match inner.version {
+            Some(
+                dapi_grpc::platform::v0::wait_for_state_transition_result_response::Version::V0(
```
Comment (QuantumExplorer) [inline packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs:327]: **Suggestion: Extract helper to reduce nesting**

All 3 tests in this file have ~20 lines of nested `match` on protobuf response enums. A helper like:
```rust
fn extract_error(response: Response<...>) -> StateTransitionBroadcastError { ... }
```
would reduce each test to 3-4 lines and make assertions more visible.

File: packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs [hunk 8]
```diff
@@ -310,3 +310,84 @@ pub(super) fn build_wait_for_state_transition_error_response(
...
+                    ),
+                ) => {
+                    // DapiError::Timeout -> to_status() -> internal with message
+                    assert!(!st_err.message.is_empty());
+                }
+                other => panic!("expected Error result, got {:?}", other),
+            },
+            other => panic!("expected V0, got {:?}", other),
+        }
+    }
```
Comment (thepastaclaw) [inline packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs:364]: **🟡 Suggestion: Weak assertion: generic error response test doesn't verify the actual code or message mapping**

The companion test `build_error_response_from_tenderdash_client_error` (line 319) asserts both `st_err.code` and `st_err.message` exactly. This test only checks `!st_err.message.is_empty()`, which passes even if the mapping changes to a completely wrong code or message.

`DapiError::Timeout` falls through to the `_` wildcard in `to_status()` (error.rs:165), producing `tonic::Status::internal(self.to_string())` — so the code is 13 (`Internal`). The message includes the `Display` output of the error. Asserting these values would document the (perhaps surprising) behavior that `Timeout` maps to `Internal` through this code path, and catch regressions.

```suggestion
                ) => {
                    // DapiError::Timeout -> to_status() -> internal with message
                    assert_eq!(st_err.code, tonic::Code::Internal as u32);
                    assert!(
                        st_err.message.contains("timed out"),
                        "expected message to contain 'timed out', got: {}",
                        st_err.message
                    );
                }
```

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs`:
- [SUGGESTION] lines 345-367: Weak assertion: generic error response test doesn't verify the actual code or message mapping
  The companion test `build_error_response_from_tenderdash_client_error` (line 319) asserts both `st_err.code` and `st_err.message` exactly. This test only checks `!st_err.message.is_empty()`, which passes even if the mapping changes to a completely wrong code or message.

`DapiError::Timeout` falls through to the `_` wildcard in `to_status()` (error.rs:165), producing `tonic::Status::internal(self.to_string())` — so the code is 13 (`Internal`). The message includes the `Display` output of the error. Asserting these values would document the (perhaps surprising) behavior that `Timeout` maps to `Internal` through this code path, and catch regressions.
```
</details>

File: packages/rs-dapi/src/metrics.rs [hunk 9]
```diff
@@ -567,3 +567,189 @@ pub fn workers_active_inc() {
...
+
+    // -- Metrics struct tests --
+
+    #[test]
+    fn metrics_struct_cache_events() {
+        let label = MethodLabel::from_type_name("test");
+        Metrics::cache_events_hit("struct_cache", &label);
+        Metrics::cache_events_miss("struct_cache", &label);
+        Metrics::cache_events_inc("struct_cache", &label, Outcome::Hit);
+    }
```
Comment (thepastaclaw) [inline packages/rs-dapi/src/metrics.rs:720]: **🟡 Suggestion: Metrics smoke tests only verify no-panic, not that metrics are recorded**

Tests like `cache_metrics_functions_do_not_panic`, `request_metrics_functions_do_not_panic`, `platform_events_metrics_do_not_panic`, and `workers_active_metrics_do_not_panic` call metric helper functions but never read back any counter/gauge values. These tests still pass if the helpers stop mutating Prometheus collectors entirely.

Consider verifying that `gather_prometheus()` output contains the expected metric names (e.g., `rsdapi_cache_events`, `rsdapi_request_count`) after exercising the helpers. This would catch regressions where a metric name or label set is misconfigured.

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dapi/src/metrics.rs`:
- [SUGGESTION] lines 596-705: Metrics smoke tests only verify no-panic, not that metrics are recorded
  Tests like `cache_metrics_functions_do_not_panic`, `request_metrics_functions_do_not_panic`, `platform_events_metrics_do_not_panic`, and `workers_active_metrics_do_not_panic` call metric helper functions but never read back any counter/gauge values. These tests still pass if the helpers stop mutating Prometheus collectors entirely.

Consider verifying that `gather_prometheus()` output contains the expected metric names (e.g., `rsdapi_cache_events`, `rsdapi_request_count`) after exercising the helpers. This would catch regressions where a metric name or label set is misconfigured.
```
</details>

File: packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs [hunk 10]
```diff
@@ -68,3 +68,236 @@ fn map_status(status: &dapi_grpc::tonic::Status) -> (i32, String, Option<Value>)
...
+        assert_eq!(msg, "Service unavailable");
+    }
+
+    #[test]
+    fn map_status_empty_message_other_code() {
+        let status = tonic::Status::new(Code::PermissionDenied, "");
+        let (_, msg, _) = map_status(&status);
+        // Other codes with empty message get "Internal error" as the normalized message
+        assert_eq!(msg, "Internal error");
+    }
```
Comment (thepastaclaw) [inline packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs:319]: **🟡 Suggestion: Missing test for `map_status` with `Code::Unimplemented`**

`map_status` has no explicit arm for `Code::Unimplemented` — it falls through to the `_` wildcard (line 64), returning `(-32603, "Internal error", Some(...))`. Meanwhile, `map_error` maps `DapiError::MethodNotFound` to `-32601`. This means a `DapiError::Status` wrapping an `Unimplemented` status maps to a *different* JSON-RPC code than `DapiError::MethodNotFound` — which could be intentional or a latent bug. Adding a test for `Code::Unimplemented` through `map_status` would document the current behavior and catch future regressions.

<sub>source: ['claude']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs`:
- [SUGGESTION] lines 293-302: Missing test for `map_status` with `Code::Unimplemented`
  `map_status` has no explicit arm for `Code::Unimplemented` — it falls through to the `_` wildcard (line 64), returning `(-32603, "Internal error", Some(...))`. Meanwhile, `map_error` maps `DapiError::MethodNotFound` to `-32601`. This means a `DapiError::Status` wrapping an `Unimplemented` status maps to a *different* JSON-RPC code than `DapiError::MethodNotFound` — which could be intentional or a latent bug. Adding a test for `Code::Unimplemented` through `map_status` would document the current behavior and catch future regressions.
```
</details>

Comment (thepastaclaw) [inline packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs:319]: **💬 Nitpick: Empty-message `map_status` tests don't verify the `data` field**

The empty-message tests (e.g., `map_status_empty_message_other_code` at line 297) only assert the `msg` return value. For codes that hit the fallback arm (lines 64-68), the `data` field is `Some(Value::String(status.to_string()))`, while for explicitly matched codes (lines 57-63) it's `None`. Asserting on `data` would make these tests more robust against regressions where the error payload is accidentally dropped or added.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

This is a well-structured test-only PR adding unit test coverage across rs-dapi. No blocking defects found. Both reviewers independently flagged the same two quality gaps: a weak assertion in the state transition error response test and smoke-only metrics tests. A few additional minor suggestions for coverage completeness.

🟡 3 suggestion(s) | 💬 1 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dapi/src/services/platform_service/wait_for_state_transition_result.rs`:
- [SUGGESTION] lines 345-367: Weak assertion: generic error response test doesn't verify the actual code or message mapping
  The companion test `build_error_response_from_tenderdash_client_error` (line 319) asserts both `st_err.code` and `st_err.message` exactly. This test only checks `!st_err.message.is_empty()`, which passes even if the mapping changes to a completely wrong code or message.

`DapiError::Timeout` falls through to the `_` wildcard in `to_status()` (error.rs:165), producing `tonic::Status::internal(self.to_string())` — so the code is 13 (`Internal`). The message includes the `Display` output of the error. Asserting these values would document the (perhaps surprising) behavior that `Timeout` maps to `Internal` through this code path, and catch regressions.

In `packages/rs-dapi/src/metrics.rs`:
- [SUGGESTION] lines 596-705: Metrics smoke tests only verify no-panic, not that metrics are recorded
  Tests like `cache_metrics_functions_do_not_panic`, `request_metrics_functions_do_not_panic`, `platform_events_metrics_do_not_panic`, and `workers_active_metrics_do_not_panic` call metric helper functions but never read back any counter/gauge values. These tests still pass if the helpers stop mutating Prometheus collectors entirely.

Consider verifying that `gather_prometheus()` output contains the expected metric names (e.g., `rsdapi_cache_events`, `rsdapi_request_count`) after exercising the helpers. This would catch regressions where a metric name or label set is misconfigured.

In `packages/rs-dapi/src/protocol/jsonrpc_translator/error.rs`:
- [SUGGESTION] lines 293-302: Missing test for `map_status` with `Code::Unimplemented`
  `map_status` has no explicit arm for `Code::Unimplemented` — it falls through to the `_` wildcard (line 64), returning `(-32603, "Internal error", Some(...))`. Meanwhile, `map_error` maps `DapiError::MethodNotFound` to `-32601`. This means a `DapiError::Status` wrapping an `Unimplemented` status maps to a *different* JSON-RPC code than `DapiError::MethodNotFound` — which could be intentional or a latent bug. Adding a test for `Code::Unimplemented` through `map_status` would document the current behavior and catch future regressions.
```
</details>
